File: src/Invoker.java
import java.util.Stack;

public class Invoker {

    private Stack<CommandCRUD> undoCommands;
    private Stack<CommandCRUD> redoCommands;
    private Middleware middleware;

    public Invoker(Middleware middleware) {
        this.undoCommands = new Stack<>();
        this.redoCommands = new Stack<>();
        this.middleware = middleware;
    }

    public void execute(CommandCRUD command){
        command.execute();
        undoCommands.push(command);
        redoCommands.clear();
    }

    public boolean undo(){
        try {
            CommandCRUD command = undoCommands.pop();
            command.unexecute();
            redoCommands.push(command);
            return true;
        }catch (Exception e){
            return false;
        }

    }

    public boolean redo(){
        try{
            CommandCRUD command = redoCommands.pop();
            command.execute();
            undoCommands.push(command);
            return true;
        }catch (Exception e){
            return false;
        }


    }

    public void commit(){
        middleware.commit();
    }

    public Stack<CommandCRUD> getUndoCommands() {
        return undoCommands;
    }

    public void setUndoCommands(Stack<CommandCRUD> undoCommands) {
        this.undoCommands = undoCommands;
    }

    public Stack<CommandCRUD> getRedoCommands() {
        return redoCommands;
    }

    public void setRedoCommands(Stack<CommandCRUD> redoCommands) {
        this.redoCommands = redoCommands;
    }

    public Middleware getMiddleware() {
        return middleware;
    }

    public void setMiddleware(Middleware middleware) {
        this.middleware = middleware;
    }
}

File: src/pattern/Invoker.js
class Invoker {
    constructor(newMiddleware) {
        this._middleware = newMiddleware;
        this._undoCommands = [];
        this._redoCommands = [];
    }
    execute(command) {
        command.execute();
        this._undoCommands.push(command);
        this._redoCommands = [];
    }
    undo() {
        try {
            let command = this._undoCommands.pop();
            command.unexecute();
            this._redoCommands.push(command);
        } catch (error) { }
        return this.getMiddleware().getCollections();
    }
    redo() {
        try {
            let command = this._redoCommands.pop();
            command.execute();
            this._undoCommands.push(command);
        } catch (error) { }
        return this.getMiddleware().getCollections();
    }
    commit() {
        this._middleware.commit();
    }
    save(type){
        this._middleware.save(type);
    }
    getMiddleware() {
        return this._middleware;
    }
    setMiddleware(newMiddleware) {
        this._middleware = newMiddleware;
    }
    getUndoCommands() {
        return this._undoCommands;
    }
    setUndoCommands(newUndoCommands) {
        this._undoCommands = newUndoCommands;
    }
    getRedoCommands() {
        return this._redoCommands;
    }
    setRedoCommands(newRedoCommands) {
        this._redoCommands = newRedoCommands;
    }
}
export default Invoker;